src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,356 @@
+
+optional arguments:
+  --root_dir ROOT_DIR   Root directory.
+  --kb_base_url KB_BASE_URL
+                        KBase base URL, defaulting to prod
+  --workers             Number of workers for multiprocessing
+  --token_filename TOKEN_FILENAME
+                        Filename in home directory that stores token
+  --delete_job_dir      Delete job directory
+
+            
+e.g.
+PYTHONPATH=. python src/loaders/workspace_downloader/workspace_downloader.py --workspace_id 39795 --kb_base_url https://ci.kbase.us/services/
+
+NOTE:
+NERSC file structure for WS:
+/global/cfs/cdirs/kbase/collections/sourcedata/ -> WS -> workspace ID -> UPA -> .fa && .meta files 
+
+e.g.
+/global/cfs/cdirs/kbase/collections/sourcedata/WS -> 39795 -> 39795_10_1 -> 39795_10_1.fa 
+                                                                         -> 39795_10_1.meta
+                                                              39795_22_1 -> 39795_22_1.fa 
+                                                                         -> 39795_22_1.meta
+                                                     
+"""
+import argparse
+import itertools
+import json
+import os
+import shutil
+import stat
+import time
+from multiprocessing import Pool, Queue, cpu_count
+
+import docker
+
+from src.clients.AssemblyUtilClient import AssemblyUtil
+from src.clients.workspaceClient import Workspace
+from src.loaders.common import loader_common_names, loader_helper
+
+# setup KB_AUTH_TOKEN as env or provide a token_filename in --token_filename
+# export KB_AUTH_TOKEN="your-kb-auth-token"
+
+# supported source of data
+SOURCE = "WS"
+# filtering applied to list objects
+FILTER_OBJECTS_NAME_BY = "KBaseGenomeAnnotations.Assembly"
+# callback_port
+CALLBACK_PORT = 9999
+
+
+class Conf:
+    def __init__(self, job_dir, output_dir, workers, kb_base_url, token_filename):
+        self.setup_callback_server_envs(job_dir, kb_base_url, token_filename)
+        time.sleep(1)
+        self.start_callback_server(docker.from_env(), "cb")
+        time.sleep(2)
+
+        # os.environ['SDK_CALLBACK_URL'] = self.cb.callback_url
+        token = os.environ["KB_AUTH_TOKEN"]
+        ws_url = os.path.join(kb_base_url, "ws")
+        callback_url = "http://" + loader_helper.get_ip() + ":" + str(CALLBACK_PORT)
+        print("callback_url:", callback_url)
+
+        self.ws = Workspace(ws_url, token=token)
+        self.asu = AssemblyUtil(callback_url, token=token)
+        self.queue = Queue()
+        self.pth = output_dir
+        self.job_dir = job_dir
+        self.pools = Pool(workers, process_input, [self])
+
+    def setup_callback_server_envs(self, job_dir, kb_base_url, token_filename):
+        # initiate env and vol
+        self.env = {}
+        self.vol = {}
+
+        # setup envs required for docker container
+        token = os.environ.get("KB_AUTH_TOKEN")
+        if not token:
+            if not token_filename:
+                raise ValueError("Need to provide a token_filename")
```

Comment:
```suggestion
                raise ValueError(f"Need to provide a token in the {names.ENV_AUTH_TOKEN} environment variable or as a file as an argument to the CLI")
```

Reply:
Updated!

Reply:
Looks good, just needs to be wrapped so it's < 100 chars

Reply:
Updated!

Reply:
:+1: